### ISSUE TITLE
lightning_logs version numbers: Include hashes of hyperparams.

### ISSUE THREAD

**Issue by Erotemic** [Lightning-AI/lightning#14739]
## 🚀 Feature

Currently lightning writes a new folder each time you start a run and it does this by incrementing a version number. It might also be nice to have the option of including a small 8-char hash of of the hyperparams. 

### Motivation

There are two motivating factors:

1. I don't like that lighting makes a new directory every time I start a run. Sometimes I just want to continue the previous run and not have to worry about multiple directories that I don't control. 

2. Tensorboard people would see which versions of their training run has the same vs different hyperparameters. 

### Pitch

Instead of getting version_0 folders when I start a training run, I want to get `version_a2d9ccj_0`, where `a2d9ccj` is a hash of the lightning hyperparameters (or at least as many of them as possible).

By default it would be good to keep the increment number as well, because its probably hard to guarantee that lightning knows about every single hparam, but there could also be an option where the user removes that and only has the hash prefix, and perhaps in that case, lightning could auto-resume a training session without needing to point at a specific path to a checkpoint. Presume that because the hparams didn't change, you want to resume from that same run and continue training in the same folder. 

### Additional context

This is a feature in my [lightning-before-lightning repo](https://github.com/Kitware/netharn) netharn, which lightning has almost entirely supplanted for me. However, there are a few features I had that don't exist in lightning, and I think this might be one of the useful ones. [Here](https://github.com/Kitware/netharn/blob/main/netharn/hyperparams.py#L826) is the part that roughly implements this 

Also I have a function [ubelt.hash_data](https://github.com/Erotemic/ubelt/blob/main/ubelt/util_hash.py#L1005) that can hash nested data structures and return hashes in arbitrary bases. For things like this I like using "abc" (i.e. base26 that only contains lowercase letters) so they aren't confused with version numbers. 

cc @borda @awaelchli @edward-io @ananthsub @rohitgr7 @Blaizzy

**Comment by awaelchli** [comment]
@Erotemic Thanks for the feature request, sorry for the late response. 

What would be your approach in integrating this feature, I assume it would be opt-in. Would every logger implementation take a new argument to control the behavior? Or was your suggestion for TensorBoard specifically? If not, what about loggers like WandB that already have a randomly generated id as the version in the folder name (a2d9ccj)? Looks similar to your hash idea but might be confusing what is what.

**Comment by Erotemic** [comment]
> randomly generated id  

The idea is that it's not random, the same hyperparams give the same id. That is, the arguments used to construct the dataset, model, and other objects would be the data fed to the hasher.

> I assume it would be opt-in. 

Yes. Opt-out would be a breaking change, and I'd like to avoid that. 

> Would every logger implementation take a new argument to control the behavior? Or was your suggestion for TensorBoard specifically?

This is not tensorboard specific, it has to do with whatever generates the folder that goes inside of "default_root_dir". I might not have the best understanding on what the mechanism that choose the incrementing versioned folders is, but checkpoints, logs, and other things need to be written somewhere and currently there seems to be something that looks for existing "version" folders and then chooses a non existing folder for (IIRC) a log_dir property. That mechanism is what I'm proposing to replace / generalize.

**Comment by awaelchli** [comment]
> The idea is that it's not random, the same hyperparams give the same id. That is, the arguments used to construct the dataset, model, and other objects would be the data fed to the hasher.

What I'm saying is that loggers like WandB will generate the random id whether we want it or not. Given your proposal, there are three choices:
A) version=<wandb-id>
B) version=<hash>
C) version=<wandb-id>_<hash>

A is the current random experiment name. It is included in the folder name so that one can easily associate the experiment and version from the online dashboard. 
B) is what you propose. However, WandB will anyway generate an experiment id in the back for it. Now, there is no easy way to associate experiments unless you look up the special hash.
C) Both are included in the folder name. This is possible but could be confusing, looking at a folder a2d9ccj_dfyb21, which one is the experiment id and which one is the special hash? 

I think we need to look at these cases to understand how to resolve this properly. I couldn't exactly see how this was managed in the linked repo. 

Note, a related work that needs to be completed first is #14188.

**Comment by Erotemic** [comment]
Having experiment_dir from #14188 might be enough for what I want. 

WRT to wandb, I was thinking that this feature would be independent of that. Let wandb do what it wants, whereas this option is mainly for vanilla lightning. What I'm looking for here is simply an alternative to `version_XXX` so I can implement a feature where if I kill a run and it has the same hyperparams it knows how to restart it automatically and it does so in the same exact folder.

Can you point me at the code where wandb choose a directory so I can understand it better?

FWIW in netharn, the concept of a log_dir is called a "train_dpath" (I use the dpath / fpath convention to differentiate directory paths and files paths). In the main harness (analogy of a trainer) there a function to [setup the paths](https://github.com/Kitware/netharn/blob/main/netharn/fit_harn.py#L493) the trainer will manage. This calls the [train_info](https://github.com/Kitware/netharn/blob/main/netharn/hyperparams.py#L849) function of the hyperparameters object, which uses the data it has to construct a "train_id", which has a bunch of different hashes representing different things. In retrospect I probably should have just used a single hash, but w/e (also in retrospect I have a huge -- and embarrassing -- error in the reasoning of one of my comments, but again w/e this repo is deprecated).

**Comment by tshu-w** [comment]
For me, the version_number of `TensorboardLogger` and `CSVLogger` is confusing as it is not easy to distinguish experiments(runs) by them.

Using the parameter hash as the bottom path (the identifier of the experiments) may be problematic because the code may have been modified, should we recover previous runs or overwrite the experiment result. However it may be convenient to save experiments with the same parameters together, and users can implement via `sub_folder` in `TensorboarderLogger` or `CSVLogger` to make the experiment_dir like `save_dir/<hash>/run_id` after https://github.com/Lightning-AI/lightning/issues/14188.

**Comment by Erotemic** [comment]
Still would be a good idea, but I suppose we can close.

**Comment by Erotemic** [comment]
@Borda perhaps in time. I have several other PRs I'm working on right now in my free time.
